lib/src/rules/prefer_enums.dart

```diff
@@ -0,0 +1,254 @@
+  * do not define `hashCode`, `==` or `index`,
+  * do not extend any class beside Object, and
+  * have no subclasses declared in the defining library.
+
+**BAD:**
+```dart
+class LogPriority {
+  static const error = LogPriority._(1, 'Error');
+  static const warning = LogPriority._(2, 'Warning');
+  static const unknown = LogPriority._unknown(2, 'Log');
+
+  final String prefix;
+  final int priorty;
+  const LogPriority._(this.priority, this.prefix);
+  const LogPriority._unknown(String prefix) : this._(-1, prefix);
+}
+```
+
+**GOOD:**
+```dart
+enum LogPriority {
+  error(1, 'Error'),
+  warning(2, 'Warning'),
+  log.unknown('Log');
+ 
+  final String prefix;
+  final int priorty;
+  LogPriority(this.priority, this.prefix);
+  LogPriority.unknown(String prefix) : this(-1, prefix);
+}
+```
+''';
+
+class PreferEnums extends LintRule {
+  PreferEnums()
+      : super(
+            name: 'prefer_enums',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    if (!context.isEnabled(Feature.enhanced_enums)) return;
+
+    // Skip contexts w/ parts.
+    if (context.allUnits.length > 1) return;
+
+    var visitor = _Visitor(this, context);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+/// A superclass for the [_EnumVisitor] and [_NonEnumVisitor].
+class _BaseVisitor extends RecursiveAstVisitor<void> {
+  /// The element representing the enum declaration that's being visited.
+  final ClassElement classElement;
+
+  _BaseVisitor(this.classElement);
+
+  /// Return `true` if the given [node] is an invocation of a generative
+  /// constructor from the class being converted.
+  bool invokesGenerativeConstructor(InstanceCreationExpression node) {
+    var constructorElement = node.constructorName.staticElement;
+    return constructorElement != null &&
+        !constructorElement.isFactory &&
+        constructorElement.enclosingElement == classElement;
+  }
+}
+
+/// A visitor used to visit the class being linted. This visitor throws an
+/// exception if a constructor for the class is invoked anywhere other than the
+/// top-level expression of an initializer for one of the fields being converted.
+///
+/// Note: copied and modified from server.
+class _EnumVisitor extends _BaseVisitor {
+  /// A flag indicating whether we are currently visiting the children of a
+  /// field declaration that will be converted to be a constant.
+  bool inConstantDeclaration = false;
+
+  List<VariableDeclaration> variableDeclarations;
+
+  _EnumVisitor(ClassElement classElement, this.variableDeclarations)
+      : super(classElement);
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (!inConstantDeclaration) {
+      if (invokesGenerativeConstructor(node)) {
+        throw _InvalidEnumException();
+      }
+    }
+    inConstantDeclaration = false;
```

Comment:
If there are multiple fields that could be constants in the same variable declaration, such as
```
static const E a = E('a'), b = E('b');
```
then this will set the flag `false` for the first field and not restore it prior to visiting the second field.

This isn't a problem in the assist because I decided to simplify the initial implementation so that we only convert the class if the fields to be converted to constants are in separate field declarations. That determination is done before we get to the visitor, so I don't have to handle it there.

This needs to be rewritten in order to support the case above to something like:
```
var wasInConstantDeclaration = inConstantDeclaration;
inConstantDeclaration = false;
super.visitInstanceCreationExpression(node);
inConstantDeclaration = wasInConstantDeclaration;
```

Reply:
I added a test case for this but it passes as is.

```dart
class A {
  static const A a = A._(), b = A._();
  const A._();
}
```

What am I missing? 🤔